### PR TITLE
Fairness and Bias Mitigation Guide

### PR DESCRIPTION
**Pull request by ArjunSubramonian**
Completed Fairness and Bias Mitigation Guide!

### CODE REVIEW

File: chapters/part3/fairness.md [hunk 1]
```diff
@@ -0,0 +1,731 @@
...
+from allennlp.common.util import prepare_environment
+from allennlp.fairness.bias_metrics import NaturalLanguageInference
+from allennlp.data import DataLoader
+from allennlp.models.archival import load_archive
+from allennlp.training.util import evaluate
+
+logger = logging.getLogger(__name__)
+
+
+@Subcommand.register("evaluate-bias-mitigation")
```
Comment (dirkgr) [inline chapters/part3/fairness.md:419]: Is this command general enough that we could just ship it?

File: chapters/part3/fairness.md [hunk 2]
```diff
@@ -0,0 +1,731 @@
...
+        "archive_file": "https://storage.googleapis.com/allennlp-public-models/snli-roberta.2021-03-11.tar.gz",
+        "module_path": "",
+        "freeze": false
+      }
+    },
+    "bias_mitigator": {
+      "type": "linear",
+      "bias_direction": {
+        "type": "two_means",
+        "seed_word_pairs_file": "https://raw.githubusercontent.com/tolga-b/debiaswe/4c3fa843ffff45115c43fe112d4283c91d225c09/data/definitional_pairs.json",
```
Comment (dirkgr) [inline chapters/part3/fairness.md:352]: Are there a lot of magical seed files? Can we "ship" them in some way, for example, but allowing the `seed_word_pairs_file` to have magical strings that map to download locations like the one you have here? That would help a lot with discoverability.

File: chapters/part3/fairness.md [hunk 3]
```diff
@@ -0,0 +1,731 @@
...
+
+Fortunately, `allennlp.fairness` provides a `Model` wrapper called [`BiasMitigatorApplicator`](http://docs.allennlp.org/main/api/fairness/bias_direction_wrappers/) to mitigate biases in contextual embeddings during finetuning on a downstream task and test time, based on Dev, S., Li, T., Phillips, J.M., & Srikumar, V. (2020). [On Measuring and Mitigating Biased Inferences of Word Embeddings](https://api.semanticscholar.org/CorpusID:201670701). ArXiv, abs/1908.09369. In this section, we detail how to apply `BiasMitigatorApplicator` to reduce binary gender bias in a large RoBERTA model for SNLI.
+
+1. The first step is to obtain the pretained model for which you would like to reduce binary gender bias. In this case, we will use [AllenNLP's public large RoBERTA model for SNLI](https://storage.googleapis.com/allennlp-public-models/snli-roberta.2021-03-11.tar.gz).
+
+2. `BiasMitigatorApplicator` works by applying a bias mitigator of your choice directly after the static embedding layer of the pretrained RoBERTA model during finetuning and evaluation. This reduces the model's ability to propagate gender biases across words in input sentences through contextualization.
```
Comment (dirkgr) [inline chapters/part3/fairness.md:313]: To me, this is the real magic, because this is the thing that allows people to use your work without having to understand it all. That's the hallmark of a good abstraction.

For that same reason, I think it's important to have a paragraph about the limitations of `BiasMitigatorApplicator`. Does it work with any model that uses transformers? Will it work equally well on all of them? Can we characterize the performance we can get out of it, both in terms of how well it mitigates bias, and in terms of the final model performance?

File: chapters/part3/fairness.md (hunk 1 above)
Comment (ArjunSubramonian) [reply chapters/part3/fairness.md:419]: It's not general enough, which is why @AkshitaB suggested that I include it in here. It could also serve as a good example for people to write their own `Subcommand`s.

File: chapters/part3/fairness.md (hunk 2 above)
Comment (ArjunSubramonian) [reply chapters/part3/fairness.md:352]: It would help with discoverability, but I'm also worried about transparency. I'm unsure if users will take the time to find the download location to which magical strings map.

File: chapters/part3/fairness.md (hunk 3 above)
Comment (ArjunSubramonian) [reply chapters/part3/fairness.md:313]: This is a great point! Just added a paragraph.

File: chapters/part3/fairness.md (hunk 2 above)
Comment (dirkgr) [reply chapters/part3/fairness.md:354]: I would expect most users to not care about the download location at all, regardless of whether we make it easy or hard to find.

Comment (ArjunSubramonian) [reply chapters/part3/fairness.md:354]: Sure, but I want users to ideally follow the link and look at what seed words are being used. I also don't like the idea of implicitly endorsing any set of seed words through magical strings; in this case, it's clear that we're just using someone else's words while not necessarily endorsing it.

File: chapters/part3/fairness.md [hunk 4]
```diff
@@ -0,0 +1,733 @@
...
+
+<em>Fairness</em> concerns ensuring that a model performs equitably for all groups of people with respect to protected attributes. For instance, we may want a model to correctly predict `[MASK]` tokens at the same rate for men, women, and non-binary genders.
+
+<em>Biases</em> are stereotypical/unjust associations a model encodes with respect to protected attributes, or subpar model performance for certain groups. Hence, <em>bias mitigation</em> is the process of lessening the severity of stereotypical/unjust associations and disparate model performance. For example, we may want to disassociate all occupations from gender, to mitigate gender bias in a downstream application that recommends jobs to individuals. 
+
+## Biases in Data
+
+We are all aware that large language models are hungry for text data, and these data greatly influence the word and sentence-level representations the models learn. NLP researchers and pratictioners often scrape data from the Internet, which include news articles, Wikipedia pages, and even patents! But, how representative are the text in these data of the real-world frequences of different words or groups of people? The answer is **not at all**.
```
Comment (schmmd) [inline chapters/part3/fairness.md:34]: `s/different words or groups of people/words among different groups of people/`

File: chapters/part3/fairness.md [hunk 5]
```diff
@@ -0,0 +1,733 @@
...
+<br>
+<br>
+This is the result of <em>reporting bias</em>, in which the "frequency with which people write about actions, outcomes, or properties is not a reflection of real-world frequencies or the degree to which a property is characteristic of a class of individuals" [2]. For instance, "murdered" occurs with a much higher frequency than does "blinked" in text corpuses because journalists usually report about murders rather than blinking, ironically as a result of the relative rarity of murders [2].
```
Comment (schmmd) [inline chapters/part3/fairness.md:37]: I would drop the word `ironically` as it's not particularly unexpected or surprising.

File: chapters/part3/fairness.md [hunk 6]
```diff
@@ -0,0 +1,733 @@
...
+
+There are many other biases in data that affect language models [2]:
+1. **Selection bias:** Data selection does not reflect a random sample, e.g. English language data are not representative of every dialect of English, Wikipedia pages overrepresent men
```
Comment (schmmd) [inline chapters/part3/fairness.md:40]: At first I was trying to understand the connection between the two examples, but eventually I realized these were two separate examples.  I might `s/are not representative of every dialect of/do not represent every dialect/`

File: chapters/part3/fairness.md [hunk 7]
```diff
@@ -0,0 +1,733 @@
...
+
+There are many other biases in data that affect language models [2]:
+1. **Selection bias:** Data selection does not reflect a random sample, e.g. English language data are not representative of every dialect of English, Wikipedia pages overrepresent men 
+
+2. **Out-group homogeneity bias:** People tend to see outgroup members as more alike than ingroup members when comparing attitudes, values, personality traits, and other characteristics, e.g. researchers incorrectly treat "non-binary" as a single, cohesive gender
+
+3. **Biased data representation:** Even with an appropriate amount of data for every group, some groups are represented less positively than others, e.g. Wikipedia pages exhibit serious biases against women
```
Comment (schmmd) [inline chapters/part3/fairness.md:44]: A citation would be helpful here.  I'm left wanting to understand this example better, but not knowing how to follow up.

File: chapters/part3/fairness.md [hunk 8]
```diff
@@ -0,0 +1,733 @@
...
+
+4. **Biased labels:** Annotations in dataset reflect the worldviews of annotators, e.g. annotators prescribe incorrect labels due to an unawareness of non-Western cultures and traditions
```
Comment (schmmd) [inline chapters/part3/fairness.md:46]: I'm sure what you're saying here is accurate, but I'm left trying to come up with an example (and the tea has not kicked in yet)--which makes me wonder whether the example could simply be explicit here.

File: chapters/part3/fairness.md [hunk 9]
```diff
@@ -0,0 +1,733 @@
...
+
+You may often hear, in the context of biases, "It's not the model, it's the data!" This is far from the truth. Models are excellent at <em>encoding</em> and <em>amplifying</em> biases. 
+
+1. **Imbalanced datasets:** Model performance inherently favors groups with rich features and a large, diverse sample, thereby marginalizing groups with limited features and a poor sample size. 
+
+2. **Spurious correlations:** Models learn complex, non-linear and spurious correlations between features and annotations that often reflect social biases and are difficult to detect and eliminate. For example, [Wang et al.](https://api.semanticscholar.org/CorpusID:195847929) discovered that even after removing subjects from images, a model could still easily predict the gender of subjects because women were photographed overwhelmingly frequently in the kitchen, that is, the model leveraged the existence of a kitchen in the image as a proxy for gender.
+
+3. **Contextualization:** LSTMs and attention-based models, through contextualization, allow for biases to be propagated across words in a sentence.
+
+Balanced datasets are not effective at taming bias amplification because of latent variables.
```
Comment (schmmd) [inline chapters/part3/fairness.md:72]: "Balanced datasets alone"?  I might provide a teaser for what else we need to do in a single sentence.  "We also need to..."

File: chapters/part3/fairness.md [hunk 10]
```diff
@@ -0,0 +1,733 @@
...
+
+## Biases in Interpretation
+
+Beyond biases in data and models, there also exist biases in interpretation [2]:
+
+1. **Confirmation bias:** The tendency to search for, interpret, favor, recall information in a way that confirms preexisting beliefs, e.g. researchers misconstrue the ability of a model to predict binary gender as the non-existence of non-binary genders
```
Comment (schmmd) [inline chapters/part3/fairness.md:78]: I'm not sure this is a great example--do you researchers actually form their world views from how model IO?  I don't really understand what it means to believe in the "non-existence of non-binary genders".  The facebook example where the model confirms the pre-existing hiring gender bias at seems like a better fit here.

File: chapters/part3/fairness.md [hunk 11]
```diff
@@ -0,0 +1,733 @@
...
+
+2. **Overgeneralization:** Coming to conclusions based on information that is too general, e.g. researchers conclude that a model works well for everyone based on its overall accuracy on data containing mostly white people
```
Comment (schmmd) [inline chapters/part3/fairness.md:80]: Facial recognition seems like a good concrete example here.

File: chapters/part3/fairness.md [hunk 12]
```diff
@@ -0,0 +1,733 @@
...
+
+3. **Correlation fallacy:** Confusing correlation with causation
+
+4. **Automation bias:** Propensity for humans to favor suggestions from automated decision-making systems over contradictory information without automation
+
+Biases in interpretation truly necessitate the creation and usage of metrics that can uncover unfairness and biased associations in models. 
+
+## Call to Action
+
+Why should we care about fairness and biases? Because ensuring fairness and mitigating biases are inherently normative [1]. As language models become more prevalent in critical decision-making systems, their unfair encoding and amplification of biases can pose serious harms for already-marginalized communities. In NLP literature, harms are often categorized into the following (Barocas and Selbst, 2016):
```
Comment (schmmd) [inline chapters/part3/fairness.md:90]: I had to look up "normative" to understand this sentence.  I'm not sure that sentence is even necessary and I might remove it, as the following sentence is clearer.  Or maybe something like "Because models are increasingly used throughout everyday life and increasingly have a real impact on all people."

File: chapters/part3/fairness.md [hunk 13]
```diff
@@ -0,0 +1,733 @@
...
+}
+```
+
+We have made the [bias-mitigated, finetuned large RoBERTA model for SNLI](https://storage.googleapis.com/allennlp-public-models/binary-gender-bias-mitigated-snli-roberta.2021-05-20.tar.gz) publicly available with a [model card](https://github.com/allenai/allennlp-models/blob/main/allennlp_models/modelcards/pair-classification-binary-gender-bias-mitigated-roberta-snli.json).
+
+</exercise>
+
+<exercise id="6" title="Next Steps">
+
+**Friendly reminder:** While the fairness and bias mitigation tools provided by `allennlp.fairness` help reduce and control biases and increase how equitably models perform, they do <em>not</em> by any means remove biases or make models entirely fair.
```
Comment (schmmd) [inline chapters/part3/fairness.md:730]: How should users react and follow up on this disclaimer?  I would reframe "by any means remove biases" to "completely remove bias".

File: chapters/part3/fairness.md [hunk 14]
```diff
@@ -312,6 +312,8 @@ Fortunately, `allennlp.fairness` provides a `Model` wrapper called [`BiasMitigat
...
 
+`BiasMitigatorApplicator` works with any Transformer model that has a static embedding layer. However, the relationships between (1) the number of epochs of finetuning, (2) the number of layers in the model, (3) the type of bias mitigator and bias direction used, and the efficacy of bias mitigation is not well-studied and a promising direction for future research. While finetuning for more epochs with the bias mitigator might push the model to depend less on gender, it could also cause the model to overfit to spurious correlations and/or discover proxies for gender that are not contained within the gender bias subspace of the mitigator. Furthermore, it is likely harder to mitigate biases for models with a larger number of layers, as they are capable of learning extremely complex mappings. And, how well a particular type of bias mitigator and bias direction work strongly depends on the finetuning data and model architecture. Fortunately, [empirical studies]((https://api.semanticscholar.org/CorpusID:201670701) have demonstrated that model performance on the evaluation set is not compromised by bias mitigation.
```
Comment (matt-gardner) [inline chapters/part3/fairness.md:315]: This is a great draft for a paragraph.  I'd recommend one change: replace the last sentence with something like the following: "You should not assume that using a BiasMitigator resolves any particular bias issue in a model.  There is published literature suggesting that these techniques should help, but that needs to be evaluated for any actual downstream use."

**Comment by schmmd** [comment]
@ArjunSubramonian here's some feedback from Noah:

```
line 16:  The tone of this part strikes me as a little off:  "This section only provides a high-level overview of fairness and how NLP models can <em>encode</em> and <em>amplify</em> biases. We highly encourage readers to consult the works of leading scholars (particularly Black women) to dive further into this field. Some excellent papers <em>in the NLP domain</em> with which to begin your journey include:"
Maybe change to:
"This section provides a high-level overview of fairness and how NLP models can <em>encode</em> and <em>amplify</em> biases.  Readers are encouraged to engage with original research on fairness and bias; we recommend the following papers:"
(the current version presumes people know nothing -- "begin your journey" -- and the idea that "leading scholars" should be privileged/prioritized seems counter to the whole spirit of fairness research.  instead of saying what attributes of authors we think readers should pay attention to, just give them the papers we recommend.  it's fine to expand the list.)
line 64:  strike "This is far from the truth." and just say "In fact, models are excellent at ..."  ("far from the truth" sounds like the kind of hyperbole that we hear in political speech -- the claim is not true, so we don't need to use exaggerated or colorful expressions to say that it's not true.)
line 90:  I agree with Michael that the sentence about "inherently normative" doesn't make sense.  "we should care about X because it is normative" is just ... circular.  a consequentialist argument will probably work better here (as in the second sentence).
line 95:  drop the exclamation mark
lines 109-113:  everywhere it says "formally" we should say "informally" -- these are not formal expressions
line 207:  I have no idea what "unlikelihood" means
```

File: chapters/part3/fairness.md (hunk 8 above)
Comment (ArjunSubramonian) [reply chapters/part3/fairness.md:46]: I will include a link to a Google AI blog that provides a great example

**Review by matt-gardner — APPROVED** [review]
I only looked at the one part that I had previously commented on, but that part LGTM.
